pool/paymentmgr.go

```diff
@@ -0,0 +1,538 @@
+func (pm *PaymentMgr) generatePayments(height uint32, coinbase dcrutil.Amount) error {
+	cfg := pm.cfg
+	switch cfg.PaymentMethod {
+	case PPS:
+		return pm.payPerShare(coinbase, height)
+
+	case PPLNS:
+		return pm.payPerLastNShares(coinbase, height)
+
+	default:
+		return fmt.Errorf("unknown payment method provided %v", cfg.PaymentMethod)
+	}
+}
+
+// isPaymentRequested asserts if a payment request exists for the
+// provided account.
+func (pm *PaymentMgr) isPaymentRequested(accountID string) bool {
+	pm.paymentReqsMtx.RLock()
+	defer pm.paymentReqsMtx.RUnlock()
+	_, ok := pm.paymentReqs[accountID]
+	return ok
+}
+
+// addPaymentRequest creates a payment request from the provided account.
+func (pm *PaymentMgr) addPaymentRequest(accountID string) {
+	pm.paymentReqsMtx.Lock()
+	pm.paymentReqs[accountID] = struct{}{}
+	pm.paymentReqsMtx.Unlock()
+}
+
+// fetchEligiblePaymentBundles fetches payment bundles greater than the
+// configured minimum payment.
+func (pm *PaymentMgr) fetchEligiblePaymentBundles(height uint32) ([]*PaymentBundle, error) {
+	maturePayments, err := fetchMaturePendingPayments(pm.cfg.DB, height)
+	if err != nil {
+		return nil, err
+	}
+	bundles := generatePaymentBundles(maturePayments)
+
+	// Iterating the bundles backwards implicitly handles decrementing the
+	// slice index when a bundle entry in the slice is removed.
+	for idx := len(bundles) - 1; idx >= 0; idx-- {
+		if bundles[idx].Total() < pm.cfg.MinPayment {
+			// Remove payments below the minimum payment if they have not been
+			// requested for by the user.
+			if !pm.isPaymentRequested(bundles[idx].Account) {
+				bundles = append(bundles[:idx], bundles[idx+1:]...)
+				continue
+			}
+			if txrules.IsDustAmount(
+				bundles[idx].Total(),
+				25, // P2PKHScriptSize
+				mempool.DefaultMinRelayTxFee) {
+				bundles = append(bundles[:idx], bundles[idx+1:]...)
+			}
+		}
+	}
+	return bundles, nil
+}
+
+// PayDividends pays mature mining rewards to participating accounts.
+func (pm *PaymentMgr) payDividends(height uint32) error {
+	// Waiting two blocks after a successful payment before proceeding with
+	// the next one because the reserved amount for transaction fees becomes
+	// change after a successful transaction. Change matures after the next
+	// block is processed. The second block is as a result of trying to
+	// maximize the transaction fee usage by processing mature payments
+	// after the transaction fees reserve has matured and ready for another
+	// transaction.
+	lastPaymentHeight := pm.fetchLastPaymentHeight()
+	if lastPaymentHeight != 0 && (height-lastPaymentHeight) < 3 {
+		return nil
+	}
+	eligiblePmts, err := pm.fetchEligiblePaymentBundles(height)
+	if err != nil {
+		return err
+	}
+	pm.paymentReqsMtx.Lock()
+	for accountID := range pm.paymentReqs {
+		delete(pm.paymentReqs, accountID)
+	}
+	pm.paymentReqsMtx.Unlock()
+
+	if len(eligiblePmts) == 0 {
+		return nil
+	}
+
+	rand.Seed(time.Now().UnixNano())
```

Comment:
I believe the function `rand.Seed` can be executed once when the process starts, I dont see any need to execute it repeatedly.

Reply:
good catch, will call it in the constructor.